building-and-packaging/material/packaging_debian_notes.md

```diff
@@ -0,0 +1,284 @@
+  ```
+
+- Build the library and install it
+
+  ```bash
+  mkdir build && cd build
+  cmake -DBUILD_SHARED_LIBS=ON ..
+  make -j
+  sudo make install
+  ```
+
+- Installs in `prefix=/usr/local`, i.e., it sets `CMAKE_INSTALL_PREFIX`.
+
+  ```bash
+  -- Install configuration: ""
+  -- Installing: /usr/local/bin/helloworld
+  -- Installing: /usr/local/lib/libsse.so
+  -- Installing: /usr/local/include/helloworld/sse.hpp
+  ```
+
+  **Note** that we have an empty configuration here. It would make sense to build the project with a proper configuration (`Debug`, `Release`...) by setting `CMAKE_BUILD_TYPE`.
+- Now we can run `helloworld` from anywhere on the system.
+    - We might need to set `PATH="/usr/local/bin":"${PATH}"`. In my container it was not set.
+- Tear down container
+
+**Note**: At this point we have a package with reasonable installation routine. This is something we would need for (different) packaging solutions such as Debian packages and
+
+## 2. Extend CMake configuration
+
+In this step we add a CPack configuration such that we can actually create a package.
+
+- We will add the configuration in another directory. Create `cmake/Packaging.cmake`
+    - This allows to separate different parts of the CMake configuration.
+- Additions to `CMakeLists.txt`
+
+  ```diff
+  + # Packaging section
+  + # Adding other cmake modules
+  + set(CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/cmake")
+  + # Include our packaging module
+  + include(Packaging)
+  ```
+
+- Edit `cmake/Packaging.cmake` and add some properties.
+    - `CPack` should be included at the end so it can ingest all preset variables
+    - We can set version, e.g., but by default it will ingest the version number from the `project()` statement in `CMakeLists.txt`.
+    - Full list of statements in [documentation](https://cmake.org/cmake/help/latest/module/CPack.html). Note, that the current statements are common properties of the packaging module.
+
+  ```diff
+  + set(CPACK_PACKAGE_NAME ${PROJECT_NAME})
+  +
+  + set(CPACK_PACKAGE_DESCRIPTION_SUMMARY "SSE hello world example project"
+  +   CACHE STRING "Extended summary.")
+  +
+  + set(CPACK_PACKAGE_VENDOR "SSE Lecturers / Employer")
+  +
+  + set(CPACK_PACKAGE_CONTACT "firstname.lastname@example.com")
+  + set(CPACK_PACKAGE_MAINTAINERS "SSE lecturers ${CPACK_PACKAGE_CONTACT}")
+  +
+  + set(CPACK_PACKAGE_HOMEPAGE_URL "https://simulation-software-engineering.github.io/homepage/")
+  +
+  + set(CPACK_PACKAGE_VERSION_MAJOR ${PROJECT_VERSION_MAJOR})
+  + set(CPACK_PACKAGE_VERSION_MINOR ${PROJECT_VERSION_MINOR})
+  + set(CPACK_PACKAGE_VERSION_PATCH ${PROJECT_VERSION_PATCH})
+  +
+  + include(CPack)
+  ```
+
+- Start up container again and then prepare project
+
+  ```bash
+  mkdir build && cd build
+  cmake -DBUILD_SHARED_LIBS=ON ..
+  ```
+
+- This will create new files `CPackConfig.cmake` and `CPackSourceConfig.cmake`
+- We can create a package, e.g., `tar.gz`
+
+  ```bash
+  cpack -G TGZ
+  ```
```

Comment:
Wouldn't `make package` do the same thing?
That's how I understood it. What did I get the wrong way?

Reply:
`make package` will create the default package(s) specified in the CPack configuration. In our case it will create three or for packages (including a `tar.gz`) package with are predefined in our case since we did not set [the `CPACK_GENERATOR` variable](https://cmake.org/cmake/help/latest/module/CPack.html#variable:CPACK_GENERATOR). `cpack` gives much more fine-grained control over what package to create.

Reply:
OK, so, if we set the correct options in the CMake file, `make package` does the trick, right?
Will you show this as well? Or too much?

Reply:
I don't have it on my list at the moment, but if you would prefer this, there is no issue adding that.

What are you using for packaging preCICE? I had the feeling that `cpack` shows up more in tutorials. I must admit that I wasn't really aware of the `package` target.